### PR TITLE
[Windows] Begin decoupling text input plugin from the view

### PR DESCRIPTION
**Pull request by loic-sharma**
Currently the text input plugin is strongly tied to a single view. This change makes the text input plugin tied to the engine in preparation for multi-view world.

Part of https://github.com/flutter/flutter/issues/115611

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: shell/platform/windows/keyboard_unittests.cc [hunk 1]
```diff
@@ -530,6 +530,14 @@ class KeyboardTest : public WindowsTest {
   KeyboardTest() = default;
   virtual ~KeyboardTest() = default;
 
+ protected:
+  void SetUp() override {
+    WindowsTest::SetUp();
+
+    // Ensure `key_calls` global is empty at the start of tests.
+    clear_key_calls();
+  }
```
Comment (loic-sharma) [inline shell/platform/windows/keyboard_unittests.cc:539]: Remove this once https://github.com/flutter/engine/pull/47829 lands

File: shell/platform/windows/text_input_plugin.h [hunk 2]
```diff
@@ -85,7 +104,7 @@ class TextInputPlugin {
   // as TextEditingDeltas or as one TextEditingValue.
   // For more information on the delta model, see:
   // https://master-api.flutter.dev/flutter/services/TextInputConfiguration/enableDeltaModel.html
-  bool enable_delta_model;
+  bool enable_delta_model = false;
```
Comment (loic-sharma) [inline shell/platform/windows/text_input_plugin.h:107]: The lack of default value caused optimized unit tests to send delta model messages unexpectedly.

File: shell/platform/windows/text_input_plugin.h [hunk 3]
```diff
@@ -72,8 +91,8 @@ class TextInputPlugin {
   // The MethodChannel used for communication with the Flutter engine.
   std::unique_ptr<flutter::MethodChannel<rapidjson::Document>> channel_;
 
-  // The associated |TextInputPluginDelegate|.
-  TextInputPluginDelegate* delegate_;
+  // The associated |FlutterWindowsEngine|.
+  FlutterWindowsEngine* engine_;
```
Comment (yaakovschectman) [inline shell/platform/windows/text_input_plugin.h:95]: Can this ever be null? Could this be a reference?

Comment (loic-sharma) [reply shell/platform/windows/text_input_plugin.h:95]: I asked @cbracken, and he confirmed this guidance also includes members passed by a constructor:

> for parameters being modified, our style is to pass as a pointer to make it clear to the reader at the call site.
> 
> Ref: https://google.github.io/styleguide/cppguide.html#Inputs_and_Outputs
> > Use non-const pointers to represent optional outputs and optional input/output parameters.

**Review by yaakovschectman — APPROVED** [review]
Alrighty then

**Comment by chinmaygarde** [comment]
Can we land this? Looks good to go.

File: shell/platform/windows/text_input_plugin.h [hunk 4]
```diff
@@ -20,33 +20,52 @@
...
+  //
+  // Triggered when the user begins editing composing text using a multi-step
+  // input method such as in CJK text input.
   virtual void ComposeBeginHook();
 
+  // Called on an IME compose commit event.
+  //
+  // Triggered when the user commits the current composing text while using a
+  // multi-step input method such as in CJK text input. Composing continues with
+  // the next keypress.
```
Comment (cbracken) [inline shell/platform/windows/text_input_plugin.h:56]: uber-nit: it's not necessarily the user who explicitly performs a commit -- in Korean, for example, just typing the next character is enough to trigger a commit of the composing region. Maybe "Called when the user triggers a commit of the [...]" to be very slightly more accurate.

File: shell/platform/windows/text_input_plugin.h [hunk 5]
```diff
@@ -20,33 +20,52 @@
...
   virtual void ComposeCommitHook();
 
+  // Called on an IME compose end event.
+  //
+  // Triggered when the user commits the composing text while using a multi-step
+  // input method such as in CJK text input.
```
Comment (cbracken) [inline shell/platform/windows/text_input_plugin.h:62]: In this case called when composing ends, which is often preceded by a commit, but if they press ESC, for example, won't be.

File: shell/platform/windows/text_input_plugin.h (hunk 3 above)
Comment (cbracken) [reply shell/platform/windows/text_input_plugin.h:95]: Yep, google c++ style dictates we use a pointer in these cases. :/

File: shell/platform/windows/text_input_plugin.h (hunk 4 above)
Comment (cbracken) [inline shell/platform/windows/text_input_plugin.h:56]: Similar comments for these comments.

File: shell/platform/windows/text_input_plugin.h (hunk 2 above)
Comment (cbracken) [reply shell/platform/windows/text_input_plugin.h:107]: Strange - this should be default zero-initialized. Do we know in which specific situations there were problems?

fwiw even if only for readability as a declaration of intent, we should make this change.

File: shell/platform/windows/text_input_plugin_unittest.cc [hunk 6]
```diff
@@ -97,21 +102,63 @@ static std::unique_ptr<rapidjson::Document> EncodedEditingState(
...
+ public:
+  TextInputPluginTest() = default;
+  virtual ~TextInputPluginTest() = default;
+
+ protected:
+  FlutterWindowsEngine* engine() { return engine_.get(); }
+  MockFlutterWindowsView* view() { return view_.get(); }
+  MockWindowBindingHandler* window() { return window_; }
+
+  void use_headless_engine() {
```
Comment (cbracken) [inline shell/platform/windows/text_input_plugin_unittest.cc:130]: Use `UpperCamelCase` for methods. For simple accessors and mutators, you can use the same naming as variables (e.g., `engine` and `set_engine`).

Ref: https://google.github.io/styleguide/cppguide.html#Function_Names

**Review by cbracken — COMMENTED** [review]
Looks good! Just a few small nits.

File: shell/platform/windows/text_input_plugin.h (hunk 4 above)
Comment (loic-sharma) [reply shell/platform/windows/text_input_plugin.h:56]: Thanks for the additional info, I'm not very familiar with this area. Updated! :)

File: shell/platform/windows/text_input_plugin.h (hunk 5 above)
Comment (loic-sharma) [reply shell/platform/windows/text_input_plugin.h:62]: Updated this!

File: shell/platform/windows/text_input_plugin_unittest.cc (hunk 6 above)
Comment (loic-sharma) [reply shell/platform/windows/text_input_plugin_unittest.cc:130]: Fixed this and the other areas where I stole this name from :)

**Review by cbracken — APPROVED** [review]
lgtm! Looks great!

**Comment by jonahwilliams** [comment]
Is this safe to land?

**Comment by loic-sharma** [comment]
Yes, landed! :)

**Pull request merged**